### PR TITLE
[PQ] Add experimental support for HPKE backed by AWS-LC KEMs

### PR DESCRIPTION
**Pull request by sgmenda-aws**
### Description of changes:

Add experimental support for using AWS-LC KEMs from the `hpke.h` interface:
- add a wrapper from `KEM` to `EVP_HPKE_KEM` and instantiate it for ML-KEM and PQ/T KEMs.
- implement the single-shot encrypt/decrypt API from the RFC.
- implement `find_by_id` for kem, kdf, and aead to simplify use from ACCP.
- update `speed.cc` to enable benchmarking of single-shot encrypt/decrypt.
- re-generate test vectors with `kem_id`

### Call-outs:

I tried to unify the `EVP_HPKE_KEM` API and the existing `KEM` API, but they are incomparable. `EVP_HPKE_KEM` does length-checking and supports authentication (`auth_encaps` and `auth_decaps` methods), while `KEM` does not. One approach to unify would be to combine `EVP_HPKE_KEM` with the KEM portion of `EVP_PKEY_METHOD` (in `p_kem.c`), but that would have required a lot of refactoring and introduce lots of merge conflicts between this experimental branch and `main`. And even if we did all that refactoring, it would only be slightly cleaner to look at, not significantly improve performance or maintainability.

### Testing:

`HPKETest` is updated to test correctness with the new added KEMs and to test the newly added single-shot and `find_by_id` APIs.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: crypto/hpke/hpke.c [hunk 1]
```diff
@@ -292,25 +308,322 @@ static int x25519_auth_decap(const EVP_HPKE_KEY *key,
...
-      x25519_auth_encap_with_seed,
-      x25519_auth_decap,
-  };
-  return &kKEM;
+  return &hpke_kem_x25519_hkdf_sha256;
+}
+
+// Secret Key Surgery
+// These stunts are performed by trained professionals, do not try this at home.
+// FIXME: This should ideally be a part of the KEM interface.
```
Comment (dkostic) [inline crypto/hpke/hpke.c:336]: you are absolutely right, I wish we had all these things nicely defined and implemented at the KEM layer -- I hope I'll get to it soon!

File: crypto/hpke/hpke.c [hunk 2]
```diff
@@ -292,25 +308,322 @@ static int x25519_auth_decap(const EVP_HPKE_KEY *key,
...
+  mlkem1024_public_from_private(public_key, secret_key);
+  OPENSSL_memcpy(
+      public_key + MLKEM1024IPD_PUBLIC_KEY_BYTES,
+      secret_key + MLKEM1024IPD_SECRET_KEY_BYTES + T384_SECRET_KEY_BYTES,
+      T384_PUBLIC_KEY_BYTES);
+}
+
+static int lckem_init_key(EVP_HPKE_KEY *key, const uint8_t *priv_key,
+                          size_t priv_key_len) {
+  const KEM *lckem = KEM_find_kem_by_nid(key->kem->internal_kem_nid);
```
Comment (dkostic) [inline crypto/hpke/hpke.c:381]: should we do NULL checks on `key`, `key->kem`, `key->private_key`, etc.?

File: crypto/hpke/hpke.c [hunk 3]
```diff
@@ -292,25 +308,322 @@ static int x25519_auth_decap(const EVP_HPKE_KEY *key,
...
+  if (lckem == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_UNSUPPORTED_ALGORITHM);
+    return 0;
+  }
+  return lckem->method->keygen(key->public_key, key->private_key);
+}
+
+// Derive a shared secret by hashing the KEM shared secret with the KEM
+// ciphertext and KEM public key. This ensures that the resulting KEM is
+// sufficiently binding.
```
Comment (dkostic) [inline crypto/hpke/hpke.c:407]: is there a reference for this?

File: crypto/hpke/hpke.c [hunk 4]
```diff
@@ -24,23 +24,38 @@
...
+#define MAX_SHARED_SECRET_LEN PQT384_SHARED_SECRET_LEN
+
+// |EVP_HPKE_KEM| API
+// - signifies that a KEM is suitable security-wise for use in |HPKE|.
+// - does buffer length checking, unlike |KEM|.
+// - supports |auth_encaps| and |auth_decaps|, unlike |KEM|.
+// - is comparable to the KEM part of |EVP_PKEY_METHOD| in p_kem.c
 struct evp_hpke_kem_st {
-  uint16_t id;
+  uint16_t id;             /* id in the HPKE spec */
```
Comment (torben-hansen) [inline crypto/hpke/hpke.c:50]:
```suggestion
  uint16_t id; // id in the HPKE spec
```
etc.

File: crypto/hpke/hpke.c [hunk 5]
```diff
@@ -292,25 +308,322 @@ static int x25519_auth_decap(const EVP_HPKE_KEY *key,
...
+  *out_enc_len = kem->enc_len;
+  return 1;
+}
+
+static int lckem_decap(const EVP_HPKE_KEY *key, uint8_t *out_shared_secret,
+                       size_t *out_shared_secret_len, const uint8_t *enc,
+                       size_t enc_len) {
+  if (enc_len != key->kem->enc_len) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_PEER_KEY);
+  }
```
Comment (torben-hansen) [inline crypto/hpke/hpke.c:476]: return error?

File: crypto/hpke/hpke.c (hunk 3 above)
Comment (sgmenda-aws) [reply crypto/hpke/hpke.c:407]: https://datatracker.ietf.org/doc/draft-connolly-cfrg-hpke-mlkem/ does this specifically for ML-KEM and it only hashes the ciphertext, but if you want to extend to all KEMs, then ideally you should hash both the public key and the ciphertext.

File: tool/speed.cc [hunk 6]
```diff
@@ -851,6 +855,139 @@ static bool SpeedKEM(std::string selected) {
...
+         SpeedSingleHPKE("HPKE-Mlkem768Sha256Chapoly",
+                         EVP_HPKE_MLKEM768_HKDF_SHA256, EVP_HPKE_HKDF_SHA256,
+                         EVP_HPKE_CHACHA20_POLY1305, selected) &&
+         SpeedSingleHPKE("HPKE-Mlkem1024Sha384Aes256gcm",
+                         EVP_HPKE_MLKEM1024_HKDF_SHA384, EVP_HPKE_HKDF_SHA384,
+                         EVP_HPKE_AES_256_GCM, selected) &&
+         SpeedSingleHPKE("HPKE-Pqt25519Sha256Aes128gcm",
+                         EVP_HPKE_PQT25519_HKDF_SHA256, EVP_HPKE_HKDF_SHA256,
+                         EVP_HPKE_AES_128_GCM, selected) &&
+         SpeedSingleHPKE("HPKE-Pqt25519768Sha256Chapoly",
```
[the remark this answers is not in the document]
Comment (sgmenda-aws) [reply tool/speed.cc]: That was a typo from regex replace.

**Review by torben-hansen — APPROVED** [review]

**Review by dkostic — APPROVED** [review]

**Pull request merged**